gtk4/src/subclass/tree_model.rs

```diff
@@ -0,0 +1,503 @@
+    iterptr: *mut ffi::GtkTreeIter,
+) {
+    let instance = &*(tree_model as *mut T::Instance);
+    let imp = instance.get_impl();
+
+    let tree: Borrowed<TreePath> = from_glib_borrow(treeptr);
+    let iter: Borrowed<TreeIter> = from_glib_borrow(iterptr);
+
+    imp.row_has_child_toggled(
+        from_glib_borrow::<_, TreeModel>(tree_model).unsafe_cast_ref(),
+        &tree,
+        &iter,
+    )
+}
+
+unsafe extern "C" fn tree_model_row_deleted<T: TreeModelImpl>(
+    tree_model: *mut ffi::GtkTreeModel,
+    treeptr: *mut ffi::GtkTreePath,
+) {
+    let instance = &*(tree_model as *mut T::Instance);
+    let imp = instance.get_impl();
+
+    let tree: Borrowed<TreePath> = from_glib_borrow(treeptr);
+
+    imp.row_deleted(
+        from_glib_borrow::<_, TreeModel>(tree_model).unsafe_cast_ref(),
+        &tree,
+    )
+}
+
+unsafe extern "C" fn tree_model_get_flags<T: TreeModelImpl>(
+    tree_model: *mut ffi::GtkTreeModel,
+) -> ffi::GtkTreeModelFlags {
+    let instance = &*(tree_model as *mut T::Instance);
+    let imp = instance.get_impl();
+
+    imp.get_flags(from_glib_borrow::<_, TreeModel>(tree_model).unsafe_cast_ref())
+        .to_glib()
+}
+
+unsafe extern "C" fn tree_model_get_n_columns<T: TreeModelImpl>(
+    tree_model: *mut ffi::GtkTreeModel,
+) -> i32 {
+    let instance = &*(tree_model as *mut T::Instance);
+    let imp = instance.get_impl();
+
+    imp.get_n_columns(from_glib_borrow::<_, TreeModel>(tree_model).unsafe_cast_ref())
+}
+
+unsafe extern "C" fn tree_model_get_column_type<T: TreeModelImpl>(
+    tree_model: *mut ffi::GtkTreeModel,
+    index: i32,
+) -> glib::ffi::GType {
+    let instance = &*(tree_model as *mut T::Instance);
+    let imp = instance.get_impl();
+
+    imp.get_column_type(
+        from_glib_borrow::<_, TreeModel>(tree_model).unsafe_cast_ref(),
+        index,
+    )
+    .to_glib()
+}
+
+unsafe extern "C" fn tree_model_get_iter<T: TreeModelImpl>(
+    tree_model: *mut ffi::GtkTreeModel,
+    iterptr: *mut ffi::GtkTreeIter,
+    pathptr: *mut ffi::GtkTreePath,
+) -> glib::ffi::gboolean {
+    let instance = &*(tree_model as *mut T::Instance);
+    let imp = instance.get_impl();
+
+    let path: Borrowed<TreePath> = from_glib_borrow(pathptr);
+
+    let ret = imp.get_iter(
+        from_glib_borrow::<_, TreeModel>(tree_model).unsafe_cast_ref(),
+        &path,
+    );
+    if let Some(iter) = ret {
+        *iterptr = *iter.to_glib_none().0;
+        true.to_glib()
+    } else {
+        false.to_glib()
+    }
+}
+
+unsafe extern "C" fn tree_model_get_path<T: TreeModelImpl>(
+    tree_model: *mut ffi::GtkTreeModel,
+    iterptr: *mut ffi::GtkTreeIter,
+) -> *mut ffi::GtkTreePath {
+    let instance = &*(tree_model as *mut T::Instance);
+    let imp = instance.get_impl();
+
+    let iter: Borrowed<TreeIter> = from_glib_borrow(iterptr);
+
+    imp.get_path(
+        from_glib_borrow::<_, TreeModel>(tree_model).unsafe_cast_ref(),
+        &iter,
+    )
+    .to_glib_full() as *mut _
```

Comment:
This misses an annotation in the C code btw

Reply:
alot of thing are not annotated in TreeModel*  :/ I will try to bring that up for gtk4.2 and hopefully fix them. The thing is I'm afraid of that nobody touched for a decade code.